File: src/main/java/me/alegian/thavma/impl/client/renderer/blockentity/AuraNodeBER.java
package me.alegian.thavma.impl.client.renderer.blockentity;

import com.mojang.blaze3d.vertex.PoseStack;
import me.alegian.thavma.impl.common.aspect.AspectMap;
import me.alegian.thavma.impl.common.block.entity.AuraNodeBE;
import me.alegian.thavma.impl.common.data.capability.AspectContainer;
import me.alegian.thavma.impl.common.data.capability.IAspectContainer;
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.MultiBufferSource;
import net.minecraft.client.renderer.RenderType;
import net.minecraft.client.renderer.blockentity.BlockEntityRenderer;
import net.minecraft.util.Mth;
import net.neoforged.api.distmarker.Dist;
import net.neoforged.api.distmarker.OnlyIn;
import net.neoforged.neoforge.client.model.data.ModelData;
import org.jetbrains.annotations.NotNull;
import org.joml.Quaternionf;

@OnlyIn(Dist.CLIENT)
public class AuraNodeBER implements BlockEntityRenderer<AuraNodeBE> {
  public static final float MIN_SCALE = 1f / 3; // containment animation minimum allowed scale
  private float scale = 1f;

  /**
   * Renders a 3x3x3 container jar around the node.
   */
  private static void renderContainer(PoseStack poseStack, @NotNull MultiBufferSource bufferSource, int combinedLight, int combinedOverlay, int containingCountdown) {
    if (containingCountdown < 0) return;

    poseStack.pushPose();
    poseStack.scale(3f, 3f, 3f);
    poseStack.translate(-0.5d, -0.5d, -0.5d);
    Minecraft.getInstance().getBlockRenderer().renderSingleBlock(T7Blocks.INSTANCE.getESSENTIA_CONTAINER().get().defaultBlockState(), poseStack, bufferSource, combinedLight, combinedOverlay, ModelData.EMPTY, RenderType.translucent());
    poseStack.popPose();
  }

  @Override
  public void render(@NotNull AuraNodeBE be, float partialTick, PoseStack poseStack, @NotNull MultiBufferSource bufferSource, int combinedLight, int combinedOverlay) {
    var containingCountdown = be.getContainingCountdown();

    poseStack.pushPose();
    this.setupPose(poseStack, containingCountdown, partialTick);

    AuraNodeBER.renderContainer(poseStack, bufferSource, combinedLight, combinedOverlay, containingCountdown);
    this.renderNode(be, poseStack, bufferSource);

    poseStack.popPose();
  }

  /**
   * Translates to the center of the block.
   * Smoothly scales down according to containment progress
   */
  private void setupPose(PoseStack poseStack, int containingCountdown, float partialTick) {
    poseStack.translate(0.5d, 0.5d, 0.5d);
    this.scale = 1f;

    if (containingCountdown < 0) return;

    this.scale = Math.max(Mth.lerp(
        partialTick,
        (float) containingCountdown / AuraNodeBE.MAX_COUNTDOWN,
        (float) (containingCountdown - 1) / AuraNodeBE.MAX_COUNTDOWN
    ), AuraNodeBER.MIN_SCALE);

    poseStack.scale(this.scale, this.scale, this.scale);
  }

  /**
   * TODO: dont forget to handle scale when using custom shader
   */
  private void renderNode(AuraNodeBE be, PoseStack poseStack, @NotNull MultiBufferSource bufferSource) {
    poseStack.pushPose();

    // follows the camera like a particle
    Quaternionf rotation = Minecraft.getInstance().gameRenderer.getMainCamera().rotation();
    poseStack.mulPose(rotation);

    AspectContainer.from(be)
        .map(IAspectContainer::getAspects)
        .map(AspectMap::toSortedList)
        .ifPresentOrElse(aspectList -> {
          for (var stack : aspectList)
            BERHelperKt.renderAuraNodeLayer(poseStack, bufferSource, stack.getAspect().getColor(), 0.4f, stack.getAmount() / 32f);
        }, () -> {
          // empty nodes look like small black circles
          BERHelperKt.renderAuraNodeLayer(poseStack, bufferSource, 0, 1, 0.5f / 32f);
        });

    poseStack.popPose();
  }
}

File: src/main/java/me/alegian/thavma/impl/client/renderer/blockentity/AuraNodeBER.kt
package me.alegian.thavma.impl.client.renderer.blockentity

import com.mojang.blaze3d.vertex.PoseStack
import me.alegian.thavma.impl.client.clientPlayerHasRevealing
import me.alegian.thavma.impl.client.util.scale
import me.alegian.thavma.impl.common.aspect.AspectMap
import me.alegian.thavma.impl.common.block.entity.AuraNodeBE
import me.alegian.thavma.impl.common.data.capability.AspectContainer
import me.alegian.thavma.impl.common.util.use
import me.alegian.thavma.impl.init.registries.deferred.T7Blocks.SEALING_JAR
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.MultiBufferSource
import net.minecraft.client.renderer.RenderType
import net.minecraft.client.renderer.blockentity.BlockEntityRenderer
import net.minecraft.util.Mth
import net.neoforged.neoforge.client.model.data.ModelData
import kotlin.math.max

private const val MAX_ALPHA = 0.4f

class AuraNodeBER : BlockEntityRenderer<AuraNodeBE> {
  private var scale = 1f

  override fun render(be: AuraNodeBE, partialTick: Float, poseStack: PoseStack, bufferSource: MultiBufferSource, combinedLight: Int, combinedOverlay: Int) {
    val containingCountdown = be.containingCountdown

    poseStack.pushPose()
    setupPose(poseStack, containingCountdown, partialTick)

    renderContainer(poseStack, bufferSource, combinedLight, combinedOverlay, containingCountdown)

    // follows the camera like a particle
    val rotation = Minecraft.getInstance().gameRenderer.mainCamera.rotation()
    poseStack.mulPose(rotation)
    AspectContainer.from(be)?.aspects?.let {
      renderNode(it, poseStack, bufferSource)
    }

    poseStack.popPose()
  }

  /**
   * Translates to the center of the block.
   * Smoothly scales down according to containment progress
   */
  private fun setupPose(poseStack: PoseStack, containingCountdown: Int, partialTick: Float) {
    poseStack.translate(0.5, 0.5, 0.5)
    scale = 1f

    if (containingCountdown < 0) return

    scale = max(
      Mth.lerp(
        partialTick,
        containingCountdown.toFloat() / AuraNodeBE.MAX_COUNTDOWN,
        (containingCountdown - 1).toFloat() / AuraNodeBE.MAX_COUNTDOWN
      ).toDouble(), MIN_SCALE.toDouble()
    ).toFloat()

    poseStack.scale(scale)
  }

  companion object {
    private const val MIN_SCALE = 1f / 3 // for lerping containment animation

    /**
     * Renders a 3x3x3 container jar around the node.
     */
    private fun renderContainer(poseStack: PoseStack, bufferSource: MultiBufferSource, combinedLight: Int, combinedOverlay: Int, containingCountdown: Int) {
      if (containingCountdown < 0) return

      poseStack.pushPose()
      poseStack.scale(3f)
      poseStack.translate(-0.5, -0.5, -0.5)
      Minecraft.getInstance().blockRenderer.renderSingleBlock(SEALING_JAR.get().defaultBlockState(), poseStack, bufferSource, combinedLight, combinedOverlay, ModelData.EMPTY, RenderType.translucent())
      poseStack.popPose()
    }

    fun renderNode(aspectMap: AspectMap, poseStack: PoseStack, bufferSource: MultiBufferSource) {
      poseStack.use {
        var alpha = 0.1f
        if (clientPlayerHasRevealing()) alpha = MAX_ALPHA

        // empty nodes look like small black circles
        aspectMap.toSortedList().run {
          if (this.isNotEmpty())
            for (stack in this)
              renderAuraNodeLayer(poseStack, bufferSource, stack.aspect.color, alpha, stack.amount / 2f / AuraNodeBE.MAX_ASPECTS)
          else renderAuraNodeLayer(poseStack, bufferSource, 0, alpha / MAX_ALPHA, 1f / AuraNodeBE.MAX_ASPECTS)
        }
      }
    }
  }
}
